### DISCUSSION TITLE
May overflow occur in some cases due to a larger vocabulary?

### DISCUSSION THREAD

**Discussion by az13js** [ggml-org/llama.cpp#1533] [General]
For example, the definition(used to return the size of vocab) in the file `llama.cpp` is:

```c++
int llama_n_vocab(const struct llama_context * ctx) {
    return ctx->vocab.id_to_token.size();
}
```

The C++ standard requires that the size of `int` be at least 16 bits, so the maximum value under the minimum size is 32767. In [ymcui/Chinese-LLaMA-Alpaca](https://github.com/ymcui/Chinese-LLaMA-Alpaca/wiki/FAQ), the vocab has been expanded to 49954. Is there an overflow that occurs when using `llama.cpp` in some cases?

```
main: build = 1 (b608b55)
main: seed  = 1684552949
llama.cpp: loading model from ggml-model-q4_0.bin
llama_model_load_internal: format     = ggjt v1 (latest)
llama_model_load_internal: n_vocab    = 49954
llama_model_load_internal: n_ctx      = 512
llama_model_load_internal: n_embd     = 4096
llama_model_load_internal: n_mult     = 256
llama_model_load_internal: n_head     = 32
llama_model_load_internal: n_layer    = 32
llama_model_load_internal: n_rot      = 128
llama_model_load_internal: ftype      = 2 (mostly Q4_0)
llama_model_load_internal: n_ff       = 11008
llama_model_load_internal: n_parts    = 1
llama_model_load_internal: model size = 7B
llama_model_load_internal: ggml ctx size =  68.20 KB
llama_model_load_internal: mem required  = 5897.00 MB (+ 1026.00 MB per state)
llama_init_from_file: kv self size  =  256.00 MB

system_info: n_threads = 4 / 8 | AVX = 1 | AVX2 = 1 | AVX512 = 1 | AVX512_VBMI = 0 | AVX512_VNNI = 0 | FMA = 1 | NEON = 0 | ARM_FMA = 0 | F16C = 1 | FP16_VA = 0 | WASM_SIMD = 0 | BLAS = 0 | SSE3 = 1 | VSX = 0 |
```

**Comment by SlyEcho** [comment]
The C standard specifies _at least_ 16 bits. In reality, since the 1990's 32-bit CPU's and OS's `int` has always been 32 bits wide.
There is an overflow, but that's because the memory requirements for different models are not calculated automatically but by hand and hardcoded, they don't account for models that have expanded the tokens that much.

This has already been reported, actually: https://github.com/ggerganov/llama.cpp/issues/1354
